spec/abci++/v3.md

```diff
@@ -0,0 +1,202 @@
+   prevote_nil = true
+  }
+  if prevote_nil {
+    broadcast ⟨PREVOTE, h_p, round_p, nil⟩
+    // Include any slashing evidence that may be sent in the process proposal response
+    for evidence in ABCI.ProcessProposal(h_p, v).evidence_list {
+      broadcast ⟨EVIDENCE, evidence⟩ 
+    }
+  }
+  step_p ← prevote
+}
+```
+
+### Prevote timeout
+
+Upon receiving 2f + 1 prevotes, setup a timeout.
+
+```go
+upon 2f + 1 ⟨PREVOTE, h_p, vr, -1⟩ 
+ with step_p = prevote for the first time, do {
+	schedule OnTimeoutPrevote(h_p, round_p) to be executed after timeoutPrevote(round_p)
+}
+```
+
+with OnTimeoutPrevote defined as:
+
+```go
+def OnTimeoutPrevote(height, round) {
+	if (height = h_p && round = round_p && step_p = prevote) {
+    precommit_extension = ABCI.ExtendVote(h_p, round_p, nil)
+		broadcast ⟨PRECOMMIT, h_p, round_p, nil, precommit_extension⟩
+		step_p ← precommit
+	}
+}
+```
+
+### Receiving enough prevotes to precommit
+
+The following code is ran upon receiving 2f + 1 prevotes for the same block
+
+```go
+upon ⟨PROPOSAL, h_p, round_p, v, -1⟩
+ from proposer(h_p, round_p)
+ AND 2f + 1 ⟨PREVOTE, h_p, vr, id(v)⟩ 
+while valid(v) ∧ step_p >= prevote for the first time do {
+ if (step_p = prevote) {
+	 lockedValue_p ← v
+	 lockedRound_p ← round_p
+   precommit_extension = ABCI.ExtendVote(h_p, round_p, id(v))
+	 broadcast ⟨PRECOMMIT, h_p, round_p, id(v), precommit_extension⟩
+   step_p ← precommit
+ }
+ validValue_p ← v
+ validRound_p ← round_p
+}
+```
+
+And upon receiving 2f + 1 prevotes for nil:
+
+```go
+upon 2f + 1 ⟨PREVOTE, h_p, round_p, nil⟩ 
+while step_p = prevote do {
+  precommit_extension = ABCI.ExtendVote(h_p, round_p, nil)
+	broadcast ⟨PRECOMMIT, h_p, round_p, nil, precommit_extension⟩
+	step_p ← precommit
+}
+```
+
+### Upon receiving a precommit
+
+Upon receiving a precommit `precommit`, we ensure that `ABCI.VerifyVoteExtension(precommit.precommit_extension) = true`
+before accepting the precommit. This is akin to how we check the signature on precommits normally, hence its not wrapped
+in the syntax of methods from the paper.
+
+### Precommit timeout
+
+Upon receiving 2f + 1 precommits, setup a timeout.
+
+```go
+upon 2f + 1 ⟨PRECOMMIT, h_p, vr, -1⟩ for the first time, do {
+	schedule OnTimeoutPrecommit(h_p, round_p) to be executed after timeoutPrecommit(round_p)
+}
+```
+
+with OnTimeoutPrecommit defined as:
+
+```go
+def OnTimeoutPrecommit(height, round) {
+	if (height = h_p && round = round_p && step_p = prevote) {
+		StartRound(round_p + 1)
+	}
+}
+```
+
+### Upon Receiving 2f + 1 precommits
+
+The following code is ran upon receiving 2f + 1 precommits for the same block
+
+```go
+upon ⟨PROPOSAL, h_p, round_p, v, -1⟩
+ from proposer(h_p, round_p)
+ AND 2f + 1 ⟨ PRECOMMIT, h_p, vr, id(v)⟩ 
+while decision_p[h_p] = nil do {
+ if (valid(v)) {
+	 decision_p[h_p] = v
+	 h_p = h_p + 1
+	 reset lockedRound_p, lockedValue_p,validRound_p and validValue_p to initial values
+	 lockedRound_p ← round_p
+   ABCI.FinalizeBlock()
```

Comment:
Again, an indentation glitch. Or is it on purpose that the ABCI calls are idented differently? 🤔